// @apollo/space-kit/ListDivider/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _tslib = require('../_tslib-bcbe0269.js');
var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
var colors = require('../colors/colors.js');

/**
 * Divider between sections in a list
 */
var ListDivider = function (props) {
    // Stop click events so we don't try to close the list when clicking something
    // non-interactive
    var handleClick = React__default.useCallback(function (event) {
        event.preventDefault();
        event.stopPropagation();
    }, []);
    return (core.jsx("hr", _tslib.__assign({}, props, { onClick: handleClick, css: core.css({
            marginLeft: -8,
            marginRight: -8,
            marginTop: 8,
            marginBottom: 8,
            borderTopColor: colors.colors.silver.base,
            borderTopWidth: 1,
            borderTopStyle: "solid",
            borderBottomWidth: 0,
            borderLeftWidth: 0,
            borderRightWidth: 0,
        }) })));
};

exports.ListDivider = ListDivider;
